File: BIG_QA_Solution/ScriptGenerator/scripts/templates/selenium_java_cucumber/src/main/java/hooks/Hooks.java
package hooks;

import io.cucumber.java.After;
import io.cucumber.java.Before;
import io.cucumber.java.Scenario;
import org.openqa.selenium.OutputType;
import org.openqa.selenium.TakesScreenshot;
import org.openqa.selenium.WebDriver;
import utils.DriverFactory;

import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;

/**
 * Cucumber lifecycle hooks that wrap a fresh WebDriver around each scenario.
 *
 * Wiring:
 *   - This class lives in the `hooks` package, which is listed in cucumber.glue (see
 *     src/test/resources/junit-platform.properties). That registration is what makes Cucumber
 *     scan for @Before / @After here.
 *   - DriverFactory owns the actual browser; this class just decides when to start/stop it
 *     and what to capture on failure. Identical shape to the Playwright/Java template's hooks.
 *
 * Failure capture:
 *   - On a failed scenario we attach the screenshot to the Cucumber report (so it shows up in
 *     cucumber.html) AND save a timestamped copy to Results/screenshots/ on disk. If the disk
 *     write fails we swallow the error — the in-report attachment is the primary artefact.
 */
public class Hooks {

    @Before
    public void beforeScenario() {
        // Fresh driver per scenario so cookies / sessionStorage / window state never leak
        // between tests.
        DriverFactory.startDriver();
    }

    @After
    public void afterScenario(Scenario scenario) {
        WebDriver driver = DriverFactory.getDriver();
        if (scenario.isFailed() && driver != null) {
            try {
                byte[] screenshot = ((TakesScreenshot) driver).getScreenshotAs(OutputType.BYTES);
                scenario.attach(screenshot, "image/png", scenario.getName());

                // Sanitise scenario name so the file path is safe on every OS.
                String safeName = scenario.getName().replaceAll("[^a-zA-Z0-9-_]", "_");
                String timestamp = LocalDateTime.now().format(DateTimeFormatter.ofPattern("yyyyMMdd_HHmmss"));
                Path dir = Paths.get("Results", "screenshots");
                Files.createDirectories(dir);
                Files.write(dir.resolve(safeName + "_" + timestamp + ".png"), screenshot);
            } catch (Exception ignored) {
                // Disk archival is best-effort; the report attachment above is sufficient.
            }
        }
        DriverFactory.quitDriver();
    }
}

File: BIG_QA_Solution/ScriptGenerator/scripts/templates/selenium_csharp_reqnroll/Hooks/Hooks.cs
using System.Text.RegularExpressions;
using OpenQA.Selenium;
using Reqnroll;
using Reqnroll.BoDi;
using SeleniumReqnrollTests.Utils;

namespace SeleniumReqnrollTests.Hooks;

/// <summary>
/// Reqnroll lifecycle hooks that wrap a fresh WebDriver around each scenario — the C# counterpart
/// of the Java template's Hooks.
///
/// Wiring:
///   - [Binding] registers this class with Reqnroll's runtime.
///   - The driver created in [BeforeScenario] is registered into the scenario's IObjectContainer
///     (Reqnroll's built-in BoDi container), so step classes receive it by simply declaring an
///     IWebDriver constructor parameter (see StepDefinitions/LoginSteps.cs). This is the idiomatic
///     Reqnroll/SpecFlow "context injection" pattern and avoids static driver state.
///
/// Failure capture: on a failed scenario we save a timestamped screenshot under
/// Results/screenshots/. Disk archival is best-effort — a failure to write it must not mask the
/// real test failure.
/// </summary>
[Binding]
public class Hooks
{
    private readonly IObjectContainer _container;
    private IWebDriver? _driver;

    // Reqnroll injects the scenario-scoped container into this constructor.
    public Hooks(IObjectContainer container)
    {
        _container = container;
    }

    [BeforeScenario]
    public void BeforeScenario()
    {
        // Fresh driver per scenario so cookies / storage / window state never leak between tests.
        var driver = DriverFactory.StartDriver();
        _driver = driver;
        // Make the driver injectable into step-definition constructors.
        _container.RegisterInstanceAs(driver);
    }

    [AfterScenario]
    public void AfterScenario(ScenarioContext scenario)
    {
        if (_driver is null) return;
        try
        {
            if (scenario.TestError is not null && _driver is ITakesScreenshot screenshotDriver)
            {
                var screenshot = screenshotDriver.GetScreenshot();
                var safeName = Regex.Replace(scenario.ScenarioInfo.Title, "[^a-zA-Z0-9-_]", "_");
                var timestamp = DateTime.Now.ToString("yyyyMMdd_HHmmss");
                var directory = Path.Combine("Results", "screenshots");
                Directory.CreateDirectory(directory);
                screenshot.SaveAsFile(Path.Combine(directory, $"{safeName}_{timestamp}.png"));
            }
        }
        catch
        {
            // Best-effort disk archival; ignore failures here.
        }
        finally
        {
            _driver.Quit();
            _driver.Dispose();
            _driver = null;
        }
    }
}
